### PR TITLE
Nj 37/38 add NJ1040 lines 6 and 7

### PR DESCRIPTION
**Pull request by jachan**
## Link to pivotal/JIRA issue
- https://github.com/newjersey/affordability-pm/issues/37
- https://github.com/newjersey/affordability-pm/issues/38

## Is PM acceptance required? (delete one)
- Yes - don't merge until JIRA issue is accepted!

## What was done?
- Added logic to the calculator, the XML submission builder, and the PDF filler to populate fields related to lines 6 and 7.
- Rebased on top of @aloverso's PDF and County Code changes, for easier PR review, switch the commits to the latest 3 commits

## How to test?
- Describe the testing approach taken to verify the changes, including:
  - Ran unit tests for the NJ classes
  - Tested with the FYST local app, using both a single filer and the Zeus one dependent persona, verified the following:
    - XML showed correct values 
    - PDF generated correctly

## Screenshots (for visual changes)
https://github.com/user-attachments/assets/fde41d3a-268b-48ed-912b-3b4658c8d353

### CODE REVIEW

File: app/lib/efile/nj/nj1040.rb [hunk 1]
```diff
@@ -3,10 +3,60 @@ module Nj
...
+      end
+
+      def calculate_line_7
+        number_of_line_7_exemptions = 0
+        if calculate_line_7_self_checkbox
+          number_of_line_7_exemptions += 1
+        end
+        if calculate_line_7_spouse_checkbox
+          number_of_line_7_exemptions += 1
+        end
```
Comment (aloverso) [inline app/lib/efile/nj/nj1040.rb:52]: [sand] This if/else structure seems pretty wordy, especially if we ever have more than two options. We're also using the same structure in `calculate_line_6`. Could we consider extracting to a private method that counts the trues and sums the total?

If you pass a list of true/false values (the checkboxes) you can for example do something like 
```
checkboxes_for_line.sum{ |a| a==true ? 1 : 0 }
```

File: app/lib/pdf_filler/nj1040_pdf.rb [hunk 2]
```diff
@@ -54,7 +54,19 @@ def hash_for_pdf
...
+
+        # line 6 exemptions
+        "Check Box39": @xml_document.at("Exemptions SpouseCuRegular")&.text == "X" ? "Yes" : "Off",
+        "Check Box40": "Off",
+        "Domestic": get_line_6_exemption_count,
+        "x  1000": get_line_6_exemption_count * 1000,
+
+        # line 7 exemptions
+        "Check Box41": @xml_document.at("Exemptions YouOver65")&.text == "X" ? "Yes" : "Off",
+        "Check Box42": @xml_document.at("Exemptions SpouseCuPartner65OrOver")&.text == "X" ? "Yes" : "Off",
```
Comment (aloverso) [inline app/lib/pdf_filler/nj1040_pdf.rb:67]: [pebble] Didn't realize the PDF value for a checkbox was "Yes" or "Off" hahaha. Can we extract this into a private utility function for getting the PDF checkbox value so that we're not repeating this ternary in many places?

File: app/lib/pdf_filler/nj1040_pdf.rb [hunk 3]
```diff
@@ -54,7 +54,19 @@ def hash_for_pdf
...
         "CountyMunicipality Code See Table page 50": @xml_document.at("ReturnHeaderState Filer USAddress CityNm")&.text,  # city / town text field
         "State": @xml_document.at("ReturnHeaderState Filer USAddress StateAbbreviationCd")&.text,
-        "ZIP Code": @xml_document.at("ReturnHeaderState Filer USAddress ZIPCd")&.text
+        "ZIP Code": @xml_document.at("ReturnHeaderState Filer USAddress ZIPCd")&.text,
+
+        # line 6 exemptions
+        "Check Box39": @xml_document.at("Exemptions SpouseCuRegular")&.text == "X" ? "Yes" : "Off",
+        "Check Box40": "Off",
+        "Domestic": get_line_6_exemption_count,
+        "x  1000": get_line_6_exemption_count * 1000,
```
Comment (aloverso) [inline app/lib/pdf_filler/nj1040_pdf.rb:63]: [pebble/dust] For this and the `x 1000_2` below, can we get this value from the XML instead of redoing the calculation here?

File: spec/lib/efile/nj/nj1040_spec.rb [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+  context 'when filing status is single' do
+    it "sets line 6 to 1000" do
+      instance.calculate
+      expect(instance.lines[:NJ1040_LINE_6].value).to eq(1000)
+    end
+
+    context 'when filer is older than 65' do
+      before do
+        intake.primary_birth_date = Date.new(1900, 1, 1)
```
Comment (aloverso) [inline spec/lib/efile/nj/nj1040_spec.rb:20]: [pebble/dust] what are the tradeoffs of changing the value on an existing intake as opposed to re-creating an intake with the desired field in the setup?

My gut says that I would always rather recreate using the factory because otherwise you open yourself up to issues regarding state leaking between tests and test order mattering to whether it passes/fails that can get really hard to debug and spot.

Was there a reason for this approach? Otherwise I'd prefer to recreate the intake rather than modify it

File: spec/factories/state_file_nj_intakes.rb [hunk 5]
```diff
@@ -110,5 +111,21 @@
...
     end
+
+    trait :married do
+      filing_status { "married_filing_jointly" }
+      spouse_birth_date { Date.new(1990, 1, 1) }
+      spouse_ssn { "123456789" }
+    end
+
+    trait :primary_over_65 do
+      primary_birth_date { Date.new(1900, 1, 1) }
```
Comment (aloverso) [inline spec/factories/state_file_nj_intakes.rb:122]: [pebble] Here and in the other tests, I would recommend using a test year that is on the edge of 65: 1959. Otherwise you're not testing the "over 65" logic that you wrote and the tests would still pass even if we accidentally flubbed our logic in the code and made it so over 55 still gets checked or something.

File: spec/lib/efile/nj/nj1040_spec.rb [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+      it 'checks the self 65+ checkbox and sets line 7 to 1000' do
+        expect(instance.lines[:NJ1040_LINE_7_SELF].value).to eq(true)
+        expect(instance.lines[:NJ1040_LINE_7_SPOUSE].value).to eq(false)
+        expect(instance.lines[:NJ1040_LINE_7].value).to eq(1000)
+      end
+    end
+
+    context 'when filer is younger than 65' do
+      before do
+        intake.primary_birth_date = Date.new(2000, 1, 1)
```
Comment (aloverso) [inline spec/lib/efile/nj/nj1040_spec.rb:32]: [pebble] same comment as above that 1960 should be the year under test for younger than 65.

File: spec/lib/efile/nj/nj1040_spec.rb [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+      end
+      it 'does not check the self 65+ checkbox and sets line 7 to 0' do
+        expect(instance.lines[:NJ1040_LINE_7_SELF].value).to eq(false)
+        expect(instance.lines[:NJ1040_LINE_7_SPOUSE].value).to eq(false)
+        expect(instance.lines[:NJ1040_LINE_7].value).to eq(0)
+      end
+    end
+  end
+
+  context 'when filing status is married filing jointly' do
```
Comment (aloverso) [inline spec/lib/efile/nj/nj1040_spec.rb:43]: this set of tests from here below is really nice & clean and well done

File: spec/lib/pdf_filler/nj1040_pdf_spec.rb [hunk 8]
```diff
@@ -92,6 +92,71 @@
...
     end
 
+    describe "exemptions" do
+      context "single filer" do
+        let(:submission) {
+          create :efile_submission, tax_return: nil, data_source: create(
+            :state_file_nj_intake,
+            primary_first_name: "Bert",
+            primary_last_name: "Muppet",
+            primary_middle_initial: "S",
```
Comment (aloverso) [inline spec/lib/pdf_filler/nj1040_pdf_spec.rb:102]: [pebble] these name lines don't seem relevant to what's being tested, can they be deleted? Here and in the tests below as well.

File: spec/lib/pdf_filler/nj1040_pdf_spec.rb [hunk 9]
```diff
@@ -92,6 +92,71 @@
...
+          expect(pdf_fields["x  1000"]).to eq "1000"
+        end
+
+        it "fills pdf with Line 7 fields" do
+          expect(pdf_fields["Check Box41"]).to eq "Off"
+          expect(pdf_fields["Check Box42"]).to eq "Off"
+          expect(pdf_fields["x  1000_2"]).to eq "0"
+          expect(pdf_fields["undefined_9"]).to eq "0"
+        end
+        context "primary over 65" do
```
Comment (aloverso) [inline spec/lib/pdf_filler/nj1040_pdf_spec.rb:117]: [nitpick] I prefer a line of whitespace in between context blocks / describe blocks / it blocks

File: spec/lib/submission_builder/ty2024/states/nj/nj_return_xml_spec.rb [hunk 10]
```diff
@@ -84,7 +78,6 @@
         it "populates line 7 XML fields" do
           expect(xml.at("Exemptions YouOver65").text).to eq("")
           expect(xml.at("Exemptions SpouseCuPartner65OrOver").text).to eq("")
-          expect(xml.at("Exemptions Line7").text).to eq("0")
```
Comment (aloverso) [inline spec/lib/submission_builder/ty2024/states/nj/nj_return_xml_spec.rb]: Why are all the calculated value tests being removed from this file?

File: app/lib/pdf_filler/nj1040_pdf.rb (hunk 3 above)
Comment (jachan) [reply app/lib/pdf_filler/nj1040_pdf.rb:63]: This value is not stored in the XML, only the calculator. Guidance from CfA was to recalculate this value here since they don't want us pulling PDF values from the calculator.

File: spec/lib/submission_builder/ty2024/states/nj/nj_return_xml_spec.rb (hunk 10 above)
Comment (jachan) [reply spec/lib/submission_builder/ty2024/states/nj/nj_return_xml_spec.rb]: Line7 and Line6 are not fields supported in the XML Schema, only the checkboxes are.

File: app/lib/efile/nj/nj1040.rb [hunk 11]
```diff
@@ -3,10 +3,60 @@ module Nj
...
       def calculate
+        set_line(:NJ1040_LINE_6_SPOUSE, :calculate_line_6_spouse_checkbox)
+        set_line(:NJ1040_LINE_6, :calculate_line_6)
+        set_line(:NJ1040_LINE_7_SELF, :calculate_line_7_self_checkbox)
+        set_line(:NJ1040_LINE_7_SPOUSE, :calculate_line_7_spouse_checkbox)
+        set_line(:NJ1040_LINE_7, :calculate_line_7)
         @lines.transform_values(&:value)
       end
 
+      def calculate_line_6_spouse_checkbox
```
[the remark this answers is not in the document]
Comment (jachan) [reply app/lib/efile/nj/nj1040.rb:19]: I'm in favor of a naming convention that clarifies true/false. I was using the `checkbox` suffix for that. I see your perspective on `calculate` being misleading, happy to simplify to `line_6_spouse_checkbox`

File: app/lib/pdf_filler/nj1040_pdf.rb (hunk 2 above)
Comment (jachan) [reply app/lib/pdf_filler/nj1040_pdf.rb:67]: Yeah "Yes" and "Off" is not my favorite either. Refactoring...

File: spec/lib/efile/nj/nj1040_spec.rb (hunk 6 above)
Comment (jachan) [reply spec/lib/efile/nj/nj1040_spec.rb:32]: Original concern was that if we chose 1957 for younger than 65, next year that would be OLDER than 65 and the test would break. I can fix this by making it dynamically generate the birthdate.

File: spec/lib/pdf_filler/nj1040_pdf_spec.rb (hunk 9 above)
Comment (jachan) [reply spec/lib/pdf_filler/nj1040_pdf_spec.rb:117]: ran rubocop, thanks for reminder

File: spec/lib/efile/nj/nj1040_spec.rb (hunk 6 above)
Comment (aloverso) [reply spec/lib/efile/nj/nj1040_spec.rb:32]: Hm. I'm not sure. It feels weird to just recreate the logic of our code in our tests (`current_year - 65` or whatever) but I also see your point about it breaking next year. 

I think I prefer the dynamic to the set-it-to-2000 approach, but I wonder if there's any other option. Hm.

File: spec/lib/efile/nj/nj1040_spec.rb (hunk 4 above)
Comment (jachan) [reply spec/lib/efile/nj/nj1040_spec.rb:20]: No reason! Didn't realize it was possible to define `primary_birth_date` in the `create(:state_file_nj_intake)` call. Thanks for demonstrating the pattern in your PR and advocating for it here.

File: spec/factories/state_file_nj_intakes.rb [hunk 12]
```diff
@@ -105,5 +111,21 @@
...
+    end
+
+    trait :married_spouse_over_65 do
```
Comment (mmazanec22) [inline spec/factories/state_file_nj_intakes.rb:125]: [dust] should this be `married_filing_jointly` or `mfj` over 65 instead so it's clearer, since there's also a married filing separately possibility?

**Review by mmazanec22 — APPROVED** [review]

File: spec/factories/state_file_nj_intakes.rb (hunk 12 above)
Comment (jachan) [reply spec/factories/state_file_nj_intakes.rb:125]: Great callout, thanks!

**Comment by jachan** [comment]
Rebased on top of PDF and County Code commits on main

File: spec/lib/submission_builder/ty2024/states/nj/nj_return_xml_spec.rb [hunk 13]
```diff
@@ -17,5 +17,77 @@
...
+
+      context "when filer is over 65" do
+        let(:intake) { create(:state_file_nj_intake, :primary_over_65) }
+        it "populates line 7 XML fields" do
+          expect(xml.at("Exemptions YouOver65").text).to eq("X")
+          expect(xml.at("Exemptions SpouseCuPartner65OrOver")).to eq(nil)
+        end
+      end
+      context "when filer is younger than 65" do
+        let(:intake) { create(:state_file_nj_intake) }
```
[the remark this answers is not in the document]
Comment (jachan) [reply spec/lib/submission_builder/ty2024/states/nj/nj_return_xml_spec.rb:40]: Thanks - was a holdover from when I was creating a trait for `:under_65`. Fixed!
